pkg/logstructured/logstructured.go

```diff
@@ -294,28 +300,57 @@ func (l *LogStructured) ttlEvents(ctx context.Context) chan *server.Event {
 				}
 			}
 		}
+
+		if ctx.Err() == nil {
+			logrus.Error("ttl events watch channel was closed due to slow processing")
+		}
 	}()
 
 	return result
 }
 
 func (l *LogStructured) ttl(ctx context.Context) {
 	// vary naive TTL support
-	mutex := &sync.Mutex{}
-	for event := range l.ttlEvents(ctx) {
-		go func(event *server.Event) {
-			select {
-			case <-ctx.Done():
-				return
-			case <-time.After(time.Duration(event.KV.Lease) * time.Second):
-			}
-			mutex.Lock()
-			if _, _, _, err := l.Delete(ctx, event.KV.Key, event.KV.ModRevision); err != nil {
-				logrus.Errorf("failed to delete expired key: %v", err)
-			}
-			mutex.Unlock()
-		}(event)
+	queue := workqueue.NewDelayingQueue()
+	go func() {
+		for l.handleTTLEvents(ctx, queue) {
+		}
+	}()
+
+	for {
+		select {
+		case <-ctx.Done():
+			queue.ShutDown()
+			return
+		default:
+		}
+
+		for event := range l.ttlEvents(ctx) {
+			queue.AddAfter(event, time.Duration(event.KV.Lease)*time.Second)
+		}
+	}
+}
+
+func (l *LogStructured) handleTTLEvents(ctx context.Context, queue workqueue.DelayingInterface) bool {
+	item, shutdown := queue.Get()
+	if shutdown {
+		logrus.Info("ttl events work queue has shut down")
```

Comment:
Log messages should generally be capitalized like a sentence. Error messages should not.
```suggestion
		logrus.Info("TTL events work queue has shut down")
```

Reply:
Fixed